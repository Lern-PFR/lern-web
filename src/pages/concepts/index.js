import ConceptDetailsPage from './ConceptDetailsPage';

export {
	// eslint-disable-next-line import/prefer-default-export
	ConceptDetailsPage,
};
