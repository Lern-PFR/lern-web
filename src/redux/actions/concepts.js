import * as ConceptsApi from 'api/conceptsApi';
import i18next from 'i18next';
import { toast } from 'react-toastify';
import { fetchSubject } from './subjects';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the concepts redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_CONCEPT_REQUEST: '@CONCEPTS/FETCH_REQUEST',
	FETCH_CONCEPT_SUCCESS: '@CONCEPTS/FETCH_SUCCESS',
	FETCH_CONCEPT_FAILURE: '@CONCEPTS/FETCH_FAILURE',

	FETCH_CONCEPT_LIST_REQUEST: '@CONCEPTS/FETCH_LIST_REQUEST',
	FETCH_CONCEPT_LIST_SUCCESS: '@CONCEPTS/FETCH_LIST_SUCCESS',
	FETCH_CONCEPT_LIST_FAILURE: '@CONCEPTS/FETCH_LIST_FAILURE',

	CREATE_CONCEPT_REQUEST: '@CONCEPTS/CREATE_REQUEST',
	CREATE_CONCEPT_SUCCESS: '@CONCEPTS/CREATE_SUCCESS',
	CREATE_CONCEPT_FAILURE: '@CONCEPTS/CREATE_FAILURE',

	UPDATE_CONCEPT_REQUEST: '@CONCEPTS/UPDATE_REQUEST',
	UPDATE_CONCEPT_SUCCESS: '@CONCEPTS/UPDATE_SUCCESS',
	UPDATE_CONCEPT_FAILURE: '@CONCEPTS/UPDATE_FAILURE',

	DELETE_CONCEPT_REQUEST: '@CONCEPTS/DELETE_REQUEST',
	DELETE_CONCEPT_SUCCESS: '@CONCEPTS/DELETE_SUCCESS',
	DELETE_CONCEPT_FAILURE: '@CONCEPTS/DELETE_FAILURE',

	CLEAR_CONCEPT_LIST: '@CONCEPTS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// //////////// Specific concept fetching actions ///////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptRequest
 * @description Action triggered anytime a concept fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_REQUEST });

/**
 * @function
 * @name fetchConceptSuccess
 * @description Action triggered as a result to a successful concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The concept retrieved from the API.
 *
 * @returns {object}
 */
const fetchConceptSuccess = ({ concept }) => ({
	type: ActionTypes.FETCH_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name fetchConceptFailure
 * @description Action triggered as a result to a failed concept fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConceptListRequest
 * @description Action triggered anytime a concept list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchConceptListRequest = () => ({ type: ActionTypes.FETCH_CONCEPT_LIST_REQUEST });

/**
 * @function
 * @name fetchConceptListSuccess
 * @description Action triggered as a result to a successful concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array}	concepts	The list of concepts retrieved from the API.
 * @param {number}	totalCount	The total amount of concepts available for the module.
 *
 * @returns {object}
 */
const fetchConceptListSuccess = ({ concepts, totalCount }) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_SUCCESS,
	payload: { concepts, totalCount },
});

/**
 * @function
 * @name fetchConceptListFailure
 * @description Action triggered as a result to a failed concept list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchConceptListFailure = (error) => ({
	type: ActionTypes.FETCH_CONCEPT_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createConceptRequest
 * @description Action triggered anytime a concept creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createConceptRequest = () => ({ type: ActionTypes.CREATE_CONCEPT_REQUEST });

/**
 * @function
 * @name createConceptSuccess
 * @description Action triggered as a result to a successfult concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The newly created concept.
 *
 * @returns {object}
 */
const createConceptSuccess = ({ concept }) => ({
	type: ActionTypes.CREATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name createConceptFailure
 * @description Action triggered as a result to a failed concept creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const createConceptFailure = (error) => ({
	type: ActionTypes.CREATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept edition actions //////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateConceptRequest
 * @description Action triggered anytime a concept edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateConceptRequest = () => ({ type: ActionTypes.UPDATE_CONCEPT_REQUEST });

/**
 * @function
 * @name updateConceptSuccess
 * @description Action triggered as a result to a successful concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The updated concept object.
 *
 * @returns {object}
 */
const updateConceptSuccess = ({ concept }) => ({
	type: ActionTypes.UPDATE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name updateConceptFailure
 * @description Action triggered as a result to a failed concept edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error The exception sent back from the API.
 *
 * @returns {object}
 */
const updateConceptFailure = (error) => ({
	type: ActionTypes.UPDATE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Concept deletion actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteConceptRequest
 * @description Action triggered anytime a concept deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteConceptRequest = () => ({ type: ActionTypes.DELETE_CONCEPT_REQUEST });

/**
 * @function
 * @name deleteConceptSuccess
 * @description Action triggered as a result to a successful concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} concept The deleted concept object.
 *
 * @returns {object}
 */
const deleteConceptSuccess = ({ concept }) => ({
	type: ActionTypes.DELETE_CONCEPT_SUCCESS,
	payload: { concept },
});

/**
 * @function
 * @name deleteConceptFailure
 * @description Action triggered as a result to a failed concept deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteConceptFailure = (error) => ({
	type: ActionTypes.DELETE_CONCEPT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Concept list clearing actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name clearConceptListAction
 * @description Action triggered anytime the concept list is not needed anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const clearConceptListAction = () => ({ type: ActionTypes.CLEAR_CONCEPT_LIST });

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchConcept
 * @description Method used to fetch the latest version of a specified concept.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 */
export const fetchConcept = (conceptId) => (dispatch) => {
	dispatch(fetchConceptRequest());

	return ConceptsApi.fetchConceptById(conceptId)
		.then((concept) => dispatch(fetchConceptSuccess({ concept })))
		.catch((error) => dispatch(fetchConceptFailure(error)));
};

/**
 * @function
 * @name fetchConceptListByModuleId
 * @description Method used to fetch all concepts from a specific module, identified by the moduleId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} moduleId The id of the module we want to retrieve concepts from.
 */
export const fetchConceptListByModuleId = (moduleId) => (dispatch) => {
	dispatch(fetchConceptListRequest());

	return ConceptsApi.fetchConceptsByModuleId(moduleId)
		.then(({ concepts, totalCount }) => dispatch(fetchConceptListSuccess({ concepts, totalCount })))
		.catch((error) => dispatch(fetchConceptListFailure(error)));
};

/**
 * @function
 * @name createConcept
 * @description Method used to create a new concept in the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData The data to create the new concept from.
 */
export const createConcept = (conceptData) => (dispatch) => {
	dispatch(createConceptRequest());

	return ConceptsApi.createConcept(conceptData)
		.then(({ concept }) => dispatch(createConceptSuccess({ concept })))
		.catch((error) => dispatch(createConceptFailure(error)));
};

/**
 * @function
 * @name updateConcept
 * @description Method used to update an existing concept instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} conceptData	The data to update the concept with.
 * @param {string} conceptId	The id to identify the concept to update.
 */
export const updateConcept = (conceptData, conceptId) => (dispatch) => {
	dispatch(updateConceptRequest());

	return ConceptsApi.updateConcept(conceptData, conceptId)
		.then(({ concept }) => dispatch(updateConceptSuccess({ concept })))
		.catch((error) => dispatch(updateConceptFailure(error)));
};

/**
 * @function
 * @name deleteConcept
 * @description Method used to remove an existing concept from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId	The id identifying the concept to remove.
 * @param {string} subjectId	The id identifying the concept's parent subject.
 */
export const deleteConcept = (conceptId, subjectId) => (dispatch) => {
	dispatch(deleteConceptRequest());

	return ConceptsApi.deleteConcept(conceptId)
		.then((concept) => {
			dispatch(deleteConceptSuccess({ concept }));
			toast.success(i18next.t('concepts.deletion.toasts.success', { name: concept.title }));
			dispatch(fetchSubject(subjectId));
		})
		.catch((error) => dispatch(deleteConceptFailure(error)));
};

/**
 * @function
 * @name clearConceptList
 * @description Method used to clear the concept list stored in memory when the application does not need it anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Promise.resolve}
 */
export const clearConceptList = () => (dispatch) => {
	dispatch(clearConceptListAction());

	return Promise.resolve();
};
