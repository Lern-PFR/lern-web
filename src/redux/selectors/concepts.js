import { sortBy } from 'lodash';
import { createSelector } from 'reselect';

/**
 * @function
 * @name getConcepts
 * @description A selector callback which returns all concepts from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getConcepts = (state) => state?.concepts?.items ?? [];

/**
 * @function
 * @name getConceptById
 * @description A selector callback which returns the concept whose id matches the conceptId parameter.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} conceptId The id of the concept we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getConceptById = createSelector(
	getConcepts,
	(_, conceptId) => conceptId,
	(concepts, conceptId) => {
		const concept = concepts.filter(({ id }) => (id === conceptId))?.[0] ?? undefined;

		if (concept === undefined) {
			return concept;
		}

		// @TODO: replace concept.courses with courses.lessons once the API has been updated.
		// Removes older versions of each lesson from the list.
		const lessons = Object.values((concept.courses || []).reduce((acc, currentLesson) => {
			const currentlyStoredLesson = acc[currentLesson.id];
			if (currentlyStoredLesson) {
				acc[currentLesson.id] = currentlyStoredLesson.version < currentLesson.version ? currentLesson : currentlyStoredLesson;
			} else {
				acc[currentLesson.id] = currentLesson;
			}

			return acc;
		}, {}));

		// @TODO: remove following snippet with the commented "return" snippet once the API has been updated.
		const { courses, ...result } = concept;

		return {
			...result,
			lessons: sortBy(lessons, 'order'),
		};

		// return {
		// 	...concept,
		//	lessons: sortBy(lessons),
		// };
	}
);

export {
	getConcepts,
	getConceptById,
};
