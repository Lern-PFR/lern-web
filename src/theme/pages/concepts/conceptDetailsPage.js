import { basmati, coconut, feedbackGreen, peppercorn, tuna } from 'theme/colors';

const pageLayout = {
	display: 'grid',
	gridTemplateColumns: '1fr 30%',
	gridGap: '32px',
	height: '100%',
	paddingTop: '40px',
	paddingBottom: '64px',
};

const sidebar = {
	position: 'sticky',
	top: '24px',
	height: '90vh',
	display: 'grid',
	padding: '24px 64px',
	borderLeft: `solid 1px ${basmati.default}`,
	gridTemplateRows: 'auto auto auto 1fr',
};

const conceptTitle = {
	margin: 0,
	lineHeight: 'initial',
	textTransform: 'capitalize',
};

const lessonContentContainer = {
	display: 'grid',
	gridTemplateRows: 'auto 1fr',
	gridGap: '24px',
};

const lessonTitleContainer = {
	position: 'sticky',
	top: 0,
	bg: coconut.default,
};

const lessonTitle = {
	margin: 0,
	lineHeight: 'initial',
	textTransform: 'capitalize',
};

const lessonDescription = {
	margin: 0,
};

const lessonContent = {
	margin: 0,
};

const questionContainer = {

};

const navigator = {
	display: 'flex',
	flexDirection: 'row',
	alignItems: 'center',
	justifyContent: 'center',
};

const navigationChevrons = {
	color: (({ disabled }) => (disabled ? tuna.lighter2 : tuna.default)),
	cursor: 'pointer',
	background: 'none',

	'&:hover': {
		color: (({ disabled }) => (disabled ? tuna.lighter2 : tuna.darker2)),
	},
};

const stepperList = {
	flex: 0.5,
	listStyle: 'none',
	padding: 0,
	display: 'flex',
	flexDirection: 'row',
	marginLeft: '1em',
	marginRight: '1em',
};

const stepper = {
	backgroundColor: basmati.default,
	flex: '1',
	height: '12px',
	border: 0,
	marginLeft: '2px',
	marginRight: '2px',
	transition: 'transform .05s ease-out, flex .05s ease-out',
	cursor: 'pointer',

	'&:hover': {
		flex: '1.3',
		transform: 'scaleY(1.3)',
		backgroundColor: peppercorn.default,
	},
};

const stepperCurrent = {
	backgroundColor: peppercorn.default,
};

const answersList = {
	listStyle: 'none',
	display: 'grid',
	gridAutoRows: '1fr',
	margin: 0,
	padding: 0,
};

const answerListItem = {
	display: 'grid',
	borderRadius: '.2em',
	gridTemplateColumns: 'min-content 1fr',
	alignItems: 'center',
	gridGap: '1em',
	padding: '1em',
	cursor: 'pointer',
};

const answerListItemInput = {
};

const validAnswerListItem = {
	backgroundColor: feedbackGreen,
};

const answerFormSubmitButton = {
	display: 'block',
	marginX: 'auto',
	marginTop: '3em',
};

const backToModuleSvg = {
	height: '.7em',
	stroke: tuna.default,
};

export {
	pageLayout,
	sidebar,
	lessonContentContainer,
	lessonTitleContainer,
	lessonTitle,
	lessonDescription,
	lessonContent,
	navigator,
	conceptTitle,
	questionContainer,
	navigationChevrons,
	stepperList,
	stepper,
	stepperCurrent,
	answersList,
	answerListItem,
	answerListItemInput,
	validAnswerListItem,
	answerFormSubmitButton,
	backToModuleSvg,
};
