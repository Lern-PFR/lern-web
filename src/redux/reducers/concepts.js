import { ActionTypes } from 'redux/actions/concepts';

const initialState = {
	isLoading: false,
	items: [],
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_CONCEPT_REQUEST:
		case ActionTypes.FETCH_CONCEPT_LIST_REQUEST:
		case ActionTypes.CREATE_CONCEPT_REQUEST:
		case ActionTypes.UPDATE_CONCEPT_REQUEST:
		case ActionTypes.DELETE_CONCEPT_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_CONCEPT_FAILURE:
		case ActionTypes.FETCH_CONCEPT_LIST_FAILURE:
		case ActionTypes.CREATE_CONCEPT_FAILURE:
		case ActionTypes.UPDATE_CONCEPT_FAILURE:
		case ActionTypes.DELETE_CONCEPT_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_CONCEPT_SUCCESS:
		case ActionTypes.DELETE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_CONCEPT_SUCCESS:
		case ActionTypes.UPDATE_CONCEPT_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [payload.concept],
				totalCount: 1,
			};
		case ActionTypes.FETCH_CONCEPT_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: [...payload.concepts],
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_CONCEPT_LIST:
			return {
				...state,
				items: [],
				totalCount: 0,
			};
		default:
			return state;
	}
};
