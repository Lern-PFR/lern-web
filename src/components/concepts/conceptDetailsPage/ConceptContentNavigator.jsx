import PropTypes from 'prop-types';
import _ from 'lodash';
import { ChevronLeft, ChevronRight } from 'react-feather';
import styled from 'styled-components';
import { StyledDiv, StyledList } from 'components/shared/styledElements';
import { navigationChevrons, navigator, stepperList } from 'theme/pages/concepts/conceptDetailsPage';

import ConceptContentNavigatorStepper from './ConceptContentNavigatorStepper';

// @TODO: Display the name of the document in a tooltip upon hovering a stepper.

const PreviousButton = styled(ChevronLeft)({ ...navigationChevrons });
const NextButton = styled(ChevronRight)({ ...navigationChevrons });

/**
 * @name ConceptContentNavigator
 * @description A component used to navigate inside a lesson's pages.
 *
 * @author Timothée Simon-Franza
 *
 * @param {number}	currentDocOrder	The current document's order value.
 * @param {array}	conceptContent	An array of the current concept's lessons and exercises.
 * @param {func}	redirectTo		Redirection method to trigger when a stepper is clicked.
 */
const ConceptContentNavigator = ({ currentDocOrder, conceptContent, redirectTo }) => {
	/**
	 * @name onPreviousClick
	 * @description onClick handler method for the 'previous' icon button.
	 *
	 * @author Timothée Simon-Franza
	 */
	const onPreviousClick = () => {
		if (currentDocOrder > 0) {
			redirectTo(currentDocOrder - 1);
		}
	};

	/**
	 * @name onNextClick
	 * @description onClick handler method for the 'next' icon button.
	 *
	 * @author Timothée Simon-Franza
	 */
	const onNextClick = () => {
		if (currentDocOrder < conceptContent.length - 1) {
			redirectTo(currentDocOrder + 1);
		}
	};

	return (
		<StyledDiv {...navigator}>
			<PreviousButton role="button" data-testid="concept-navigation-previous" onClick={onPreviousClick} disabled={currentDocOrder === 0} background="white" />
			<StyledList {...stepperList}>
				{conceptContent.map(({ id, order, title = '', name = '' }) => (
					<ConceptContentNavigatorStepper
						isCurrent={_.isEqual(conceptContent[currentDocOrder]?.id, id)}
						key={id}
						data-testid={`concept-navigation-stepper-${id}`}
						label={title || name}
						onClick={() => redirectTo(order)}
					/>
				))}
			</StyledList>
			<NextButton role="button" data-testid="concept-navigation-next" onClick={onNextClick} disabled={currentDocOrder === (conceptContent.length - 1)} />
		</StyledDiv>
	);
};

ConceptContentNavigator.propTypes = {
	currentDocOrder: PropTypes.number.isRequired,
	conceptContent: PropTypes.arrayOf(
		PropTypes.oneOfType([
			PropTypes.shape({
				id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
				title: PropTypes.string,
				name: PropTypes.string,
				order: PropTypes.number.isRequired,
			}),
			PropTypes.shape({
				id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
				name: PropTypes.string.isRequired,
				order: PropTypes.number.isRequired,
			}),
		])
	).isRequired,
	redirectTo: PropTypes.func.isRequired,
};

export default ConceptContentNavigator;
