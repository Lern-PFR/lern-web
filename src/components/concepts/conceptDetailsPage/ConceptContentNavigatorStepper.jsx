import PropTypes from 'prop-types';
import styled from 'styled-components';
import { stepper, stepperCurrent } from 'theme/pages/concepts/conceptDetailsPage';

const StepperButton = styled('button')(
	{
		...stepper,
		backgroundColor: (({ isCurrent }) => (isCurrent ? stepperCurrent.backgroundColor : stepper.backgroundColor)),
	}
);

/**
 * @name ConceptContentNavigatorStepper
 * @description A stepper button used to navigate between a concept's content list.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {bool}	isCurrent	Whether the stepper represents the currently displayed document.
 * @param {string}	label		The label to display on hover.
 * @param {func}	onClick		Method to trigger when the user clicks the button.
 */
const ConceptContentNavigatorStepper = ({ isCurrent, label, onClick, ...otherProps }) => (
	<StepperButton onClick={onClick} label={label} isCurrent={isCurrent} {...otherProps} />
);

ConceptContentNavigatorStepper.propTypes = {
	isCurrent: PropTypes.bool,
	label: PropTypes.string.isRequired,
	onClick: PropTypes.func.isRequired,
};

ConceptContentNavigatorStepper.defaultProps = {
	isCurrent: false,
};

export default ConceptContentNavigatorStepper;
