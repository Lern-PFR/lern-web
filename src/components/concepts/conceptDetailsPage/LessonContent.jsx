import { StyledDiv } from 'components/shared/styledElements';
import { BodyCopy, LongPrimer, Trafalgar } from 'components/shared/typography';
import PropTypes from 'prop-types';
import Editor from 'rich-markdown-editor';
import { lessonContent, lessonContentContainer, lessonDescription, lessonTitle, lessonTitleContainer } from 'theme/pages/concepts/conceptDetailsPage';

/**
 * @name LessonContent
 * @description A component used to display a lesson's content.
 *
 * @author TimothÃ©e Simon-Franza
 * @author Yann Hodiesne
 *
 * @param {string} title		The current lesson's title.
 * @param {string} content		The current lesson's content.
 * @param {string} description	A short description of the current lesson.
 * @param {string} [question]	The question's statement, if any.
 */
const LessonContent = ({ title, content, description, question }) => (
	<StyledDiv {...lessonContentContainer}>
		<StyledDiv {...lessonTitleContainer}>
			<Trafalgar {...lessonTitle} tag="h2">{title}</Trafalgar>
			<LongPrimer {...lessonDescription}>{description}</LongPrimer>
		</StyledDiv>
		<StyledDiv>
			<BodyCopy tag="div" {...lessonContent}>
				<Editor defaultValue={content} value={content} readOnly />
			</BodyCopy>
			{question && (
				<BodyCopy>
					<strong>Question : </strong>
					{question}
				</BodyCopy>
			)}
		</StyledDiv>
	</StyledDiv>
);

LessonContent.propTypes = {
	title: PropTypes.string.isRequired,
	content: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	question: PropTypes.string,
};

LessonContent.defaultProps = {
	question: null,
};

export default LessonContent;
