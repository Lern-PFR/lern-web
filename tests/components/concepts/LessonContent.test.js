import { shallow } from 'enzyme';
import { LessonContent } from 'components/concepts/conceptDetailsPage';

describe('LessonContent', () => {
	const lessonData = {
		title: 'test lesson',
		content: 'test lesson content',
		description: 'test lesson description',
	};

	it('should match previous snapshot (no question provided)', () => {
		const wrapper = shallow(<LessonContent {...lessonData} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should match previous snapshot (with question provided)', () => {
		const wrapper = shallow(<LessonContent {...lessonData} question="test lesson question" />);
		expect(wrapper).toMatchSnapshot();
	});
});
