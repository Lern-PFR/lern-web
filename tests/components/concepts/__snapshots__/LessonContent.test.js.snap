// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonContent should match previous snapshot (no question provided) 1`] = `
<styled.div
  display="grid"
  gridGap="24px"
  gridTemplateRows="auto 1fr"
>
  <styled.div
    bg="#FFFFFF"
    position="sticky"
    top={0}
  >
    <Trafalgar
      lineHeight="initial"
      margin={0}
      tag="h2"
      textTransform="capitalize"
    >
      test lesson
    </Trafalgar>
    <LongPrimer
      margin={0}
    >
      test lesson description
    </LongPrimer>
  </styled.div>
  <styled.div>
    <BodyCopy
      margin={0}
      tag="div"
    >
      <RichMarkdownEditor
        defaultValue="test lesson content"
        dir="auto"
        embeds={Array []}
        extensions={Array []}
        onClickLink={[Function]}
        onImageUploadStart={[Function]}
        onImageUploadStop={[Function]}
        placeholder="Write something nice…"
        readOnly={true}
        tooltip={[Function]}
        value="test lesson content"
      />
    </BodyCopy>
  </styled.div>
</styled.div>
`;

exports[`LessonContent should match previous snapshot (with question provided) 1`] = `
<styled.div
  display="grid"
  gridGap="24px"
  gridTemplateRows="auto 1fr"
>
  <styled.div
    bg="#FFFFFF"
    position="sticky"
    top={0}
  >
    <Trafalgar
      lineHeight="initial"
      margin={0}
      tag="h2"
      textTransform="capitalize"
    >
      test lesson
    </Trafalgar>
    <LongPrimer
      margin={0}
    >
      test lesson description
    </LongPrimer>
  </styled.div>
  <styled.div>
    <BodyCopy
      margin={0}
      tag="div"
    >
      <RichMarkdownEditor
        defaultValue="test lesson content"
        dir="auto"
        embeds={Array []}
        extensions={Array []}
        onClickLink={[Function]}
        onImageUploadStart={[Function]}
        onImageUploadStop={[Function]}
        placeholder="Write something nice…"
        readOnly={true}
        tooltip={[Function]}
        value="test lesson content"
      />
    </BodyCopy>
    <BodyCopy>
      <strong>
        Question : 
      </strong>
      test lesson question
    </BodyCopy>
  </styled.div>
</styled.div>
`;
